### PR TITLE
Document distribution of Kedro pipelines with Dask

### PR DESCRIPTION
**Pull request by deepyaman**
## Description

Update #1131, to which I no longer have write access. This can be either merged directly or merged into that PR, which can then be merged.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1248"></a>

### CODE REVIEW

File: RELEASE.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## Major features and improvements
 * `pipeline` now accepts `tags` and a collection of `Node`s and/or `Pipeline`s rather than just a single `Pipeline` object. `pipeline` should be used in preference to `Pipeline` when creating a Kedro pipeline.
 * `pandas.SQLTableDataSet` and `pandas.SQLQueryDataSet` now only open one connection per database, at instantiation time (therefore at catalog creation time), rather than one per load/save operation.
+* Documented distribution of Kedro pipelines with Dask.
```
Comment (lorenabalan) [inline RELEASE.md:6]: Apologies, this will have to move to a new Release header.

File: RELEASE.md [hunk 2]
```diff
@@ -406,7 +407,7 @@ Check your source directory. If you defined a different source directory (`sourc
 
 ## Major features and improvements
 
-* Added documentation with a focus on single machine and distributed environment deployment; the series includes Docker, Argo, Prefect, Kubeflow, AWS Batch, AWS Sagemaker and extends our section on Databricks
+* Added documentation with a focus on single machine and distributed environment deployment; the series includes Docker, Argo, Prefect, Kubeflow, AWS Batch, AWS Sagemaker and extends our section on Databricks.
```
Comment (lorenabalan) [inline RELEASE.md:410]: 😂

File: docs/source/03_tutorial/05_visualise_pipeline.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ kedro viz
 ```
 
 This command will run a server on http://127.0.0.1:4141 that will open up your visualisation on a browser. You should
- be able to see the following:
+be able to see the following:
```
Comment (lorenabalan) [inline docs/source/03_tutorial/05_visualise_pipeline.md:20]: can we actually just move it on the same line? We don't break up long lines in our .md files, we just use wrap-around in IDE.

File: docs/source/index.rst [hunk 4]
```diff
@@ -150,6 +150,7 @@ Welcome to Kedro's documentation!
    10_deployment/09_aws_sagemaker
    10_deployment/10_aws_step_functions
    10_deployment/11_airflow_astronomer
+   10_deployment/12_dask
```
Comment (lorenabalan) [inline docs/source/index.rst:153]: Can we actually rename this page to just `dask`? :) On `develop` we got rid of numbering in our docs, so that we don't break links as often, and this would be one less conflict to resolve on `merge-main-to-develop`.

File: docs/source/10_deployment/12_dask.md [hunk 5]
```diff
@@ -0,0 +1,313 @@
+# Deployment to a Dask cluster
+
+This page explains how to distribute execution of the nodes composing your Kedro pipeline using [Dask](https://docs.dask.org/en/stable/), a flexible, open-source library for parallel computing in Python.
+
+Dask offers both a default, single-machine scheduler and a more sophisticated, distributed scheduler. The newer [`dask.distributed`](http://distributed.dask.org/en/stable/) scheduler is often preferable, even on single workstations, and is the focus of our deployment guide. For more information on the multitude of ways to set up Dask on varied hardware, see [the official how-to guide](https://docs.dask.org/en/stable/how-to/deploy-dask-clusters.html).
```
Comment (lorenabalan) [inline docs/source/10_deployment/12_dask.md:5]:
```suggestion
Dask offers both a default, single-machine scheduler and a more sophisticated, distributed scheduler. The newer [`dask.distributed`](http://distributed.dask.org/en/stable/) scheduler is often preferable, even on single workstations, and is the focus of our deployment guide. For more information on the various ways to set up Dask on varied hardware, see [the official how-to guide](https://docs.dask.org/en/stable/how-to/deploy-dask-clusters.html).
```

File: docs/source/10_deployment/12_dask.md [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+## Why would you use Dask?
+
+`Dask.distributed` is a lightweight library for distributed computing in Python. It complements the existing PyData analysis stack, which forms the basis of many Kedro pipelines. It's also pure Python, which eases installation and simplifies debugging. For further motivation on why people choose to adopt Dask, and, more specifically, `dask.distributed`, see [Why Dask?](https://docs.dask.org/en/stable/why.html) and [the `dask.distributed` docs](http://distributed.dask.org/en/stable/#motivation), respectively.
```
Comment (lorenabalan) [inline docs/source/10_deployment/12_dask.md:9]:
```suggestion
`Dask.distributed` is a lightweight library for distributed computing in Python. It complements the existing PyData analysis stack, which forms the basis of many Kedro pipelines. It's also pure Python, which eases installation and simplifies debugging. For further motivation on why people choose to adopt Dask, and, more specifically, `dask.distributed`, see [Why Dask?](https://docs.dask.org/en/stable/why.html) and [the `dask.distributed` documentation](http://distributed.dask.org/en/stable/#motivation), respectively.
```

File: docs/source/10_deployment/12_dask.md [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+
+where the helper function `_instantiate_runner()` looks like this:
+
+```python
+def _instantiate_runner(runner, is_async, project_context):
+    runner_class = load_obj(runner, "kedro.runner")
+    runner_kwargs = dict(is_async=is_async)
+
+    if runner.endswith("DaskRunner"):
+        client_args = project_context.params.get("dask_client") or {}
```
Comment (lorenabalan) [inline docs/source/10_deployment/12_dask.md:267]: Wouldn't this be 
```suggestion
        client_args = project_context.params.get("params.dask_client") or {}
```
?

File: docs/source/10_deployment/12_dask.md [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        Args:
+            ds_name: Name of the missing dataset.
+
+        Returns:
+            An instance of ``_DaskDataSet`` to be used for all
+            unregistered datasets.
+        """
+        return _DaskDataSet(ds_name)
+
+    def _run_node(
```
Comment (lorenabalan) [inline docs/source/10_deployment/12_dask.md:110]: Does this need to be a class method, or can it be a standalone function? As far as I can tell it's just a thin wrapper around `run_node` so that you pass `dependencies`, but they're not actually used here. Presumable the dask client is?

**Review by lorenabalan — APPROVED** [review]

**Comment by lorenabalan** [comment]
LGTM, thanks for tidying up some other parts of the docs while here! 🙌

File: docs/source/index.rst (hunk 4 above)
Comment (deepyaman) [reply docs/source/index.rst:153]: Done!

File: RELEASE.md (hunk 1 above)
Comment (deepyaman) [reply RELEASE.md]: Done!

File: docs/source/10_deployment/12_dask.md (hunk 7 above)
Comment (deepyaman) [reply docs/source/10_deployment/12_dask.md]: Don't think so:

```
(kedro-dask-example) WASH-178551-C02X31K9JHD4:kedro-dask-example deepyaman_datta$ kedro run --env dask --runner kedro_dask_example.runner.DaskRunner
2022-02-24 22:03:44,996 - kedro.framework.cli.hooks.manager - INFO - Registered CLI hooks from 1 installed plugin(s): kedro-telemetry-0.1.3
2022-02-24 22:03:45,179 - kedro_telemetry.plugin - INFO - You have opted into product usage analytics.
2022-02-24 22:03:45,492 - kedro.framework.session.store - INFO - `read()` not implemented for `BaseSessionStore`. Assuming empty store.
> /Users/deepyaman_datta/kedro-dask-example/src/kedro_dask_example/cli.py(61)_instantiate_runner()
-> client_args = project_context.params.get("dask_client") or {}
(Pdb) project_context.params
{'example_test_data_ratio': 0.2, 'example_num_train_iter': 10000, 'example_learning_rate': 0.01, 'dask_client': {'address': '127.0.0.1:8786'}}
(Pdb) 
```

This is the same way I've been running github.com/deepyaman/kedro-dask-example.

File: docs/source/10_deployment/12_dask.md (hunk 8 above)
Comment (deepyaman) [reply docs/source/10_deployment/12_dask.md]: @lorenabalan It doesn't need to be an instance method. I could make it into a static method, if so desired?  It could also be outside the class, but I think:

1. grouping for functionality is valid reason to make it into a static method
2. for the purpose of a deployment guide, having something outside the class is harder to keep track of (not really, but arguably)

Open to doing whatever with this; let me know if you have a strong preference. I can leave it as is or make it into a static method, with least preference for moving it outside the class, but can roll with anything.

Comment (lorenabalan) [reply docs/source/10_deployment/12_dask.md]: Yeah fair enough about the grouping aspect. Let's stick a `staticmethod` for this one.

File: docs/source/10_deployment/12_dask.md (hunk 7 above)
Comment (lorenabalan) [reply docs/source/10_deployment/12_dask.md]: Oh interesting, good to be reminded of that. 😂  Thanks for checking!

File: docs/source/10_deployment/dask.md [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+            ds_name: Name of the missing dataset.
+
+        Returns:
+            An instance of ``_DaskDataSet`` to be used for all
+            unregistered datasets.
+        """
+        return _DaskDataSet(ds_name)
+
+    def _run_node(
+        self,
```
Comment (deepyaman) [inline docs/source/10_deployment/dask.md:111]:
```suggestion
    @staticmethod
    def _run_node(
```

File: docs/source/10_deployment/dask.md [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+        node_futures = {}
+
+        client = Client.current()
+        for node in nodes:
+            dependencies = (
+                node_futures[dependency] for dependency in node_dependencies[node]
+            )
+            node_futures[node] = client.submit(
+                self._run_node, node, catalog, self._is_async, run_id, *dependencies
+            )
```
Comment (deepyaman) [inline docs/source/10_deployment/dask.md:153]:
```suggestion
                DaskRunner._run_node,
                node,
                catalog,
                self._is_async,
                run_id,
                *dependencies,
            )
```

File: docs/source/10_deployment/12_dask.md (hunk 8 above)
Comment (deepyaman) [reply docs/source/10_deployment/12_dask.md]: Done!

**Review by AntonyMilneQB — APPROVED** [review]
This is really awesome stuff, thank you very much @deepyaman. 

Very happy for this to be merged - the questions I ask here and in the inline comments are just to satisfy my curiosity or contribute to more general ponderings.

I know this follows the same pattern as the AWS batch runner, which is fine, but there's a couple of things that definitely seem hacky about it (as they do for the AWS batch runner). I don't immediately see any better way of doing it now, but I wonder what it would take on the kedro side to make this sort of extension more elegant in future.
1. Injecting custom arguments into `DaskRunner` instantiation via conf/dask/parameters.yml. This is very cunning, but doesn't seem ideal because it's incompatible with running any other run configuration environment (unless you also modify the ConfigLoader to allow for multiple environments to load on top of each other).
2. `def run` is essentially exactly the same as the default one; the only difference (unless I'm missing something?) is that it enables you to pass custom arguments into the runner instantiation.

I wonder if there's some other way of configuring our `run` command in order to make it easier to do this sort of thing. This issue seems very relevant, and I'd be interested if you had any thoughts to add to it based on what you've seen here: https://github.com/kedro-org/kedro/issues/1041

Naively it seems like we could expose `RUNNER_CLASS` and `RUNNER_CLASS_ARGS` in settings.py to enable this sort of thing more directly. But given that the runner is arguably runtime configuration (that belongs in conf) rather than application settings (that belongs in settings.py) that probably doesn't make sense. Sooo I don't know how we can support custom runner + arguments a bit more naturally.

File: docs/source/10_deployment/dask.md [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+            # for the shared, default datasets we created above.
+            for data_set in node.inputs:
+                load_counts[data_set] -= 1
+                if load_counts[data_set] < 1 and data_set not in pipeline.inputs():
+                    catalog.release(data_set)
+            for data_set in node.outputs:
+                if load_counts[data_set] < 1 and data_set not in pipeline.outputs():
+                    catalog.release(data_set)
+
+    def run_only_missing(
```
[the remark this answers is not in the document]
Comment (deepyaman) [reply docs/source/10_deployment/dask.md:177]: 1. I implemented this because I felt it was especially relevant for Dask. The `DaskDataSet` actually publishes data to the cluster, so there's less progress lost than in the case of `MemoryDataSet` in case of an error.
2. I did test it months ago when I wrote the behavior, but I probably hacked it in somewhere for testing. Not aware of an easy call off the top of my head (but it probably should be!). Looks like it was raised a few years ago (#30, #60) but decided against back then.

File: docs/source/10_deployment/dask.md [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+
+Next, [set up scheduler and worker processes on your local computer](http://distributed.dask.org/en/stable/quickstart.html#setup-dask-distributed-the-hard-way):
+
+```bash
+$ dask-scheduler
+Scheduler started at 127.0.0.1:8786
+
+$ PYTHONPATH=$PWD/src dask-worker 127.0.0.1:8786
+$ PYTHONPATH=$PWD/src dask-worker 127.0.0.1:8786
+$ PYTHONPATH=$PWD/src dask-worker 127.0.0.1:8786
```
[the remark this answers is not in the document]
Comment (deepyaman) [reply docs/source/10_deployment/dask.md:303]: You need to somehow make the code available to the worker, and in case of a single-machine scheduler, this works.

`Client.upload_file` is cleaner, and something like that would be necessary for distributed deployment.

https://stackoverflow.com/a/39994128/1093967 for some more details.

**Comment by deepyaman** [comment]
> P.S. fun fact, don't know if you're aware of the [recent change](https://github.com/kedro-org/kedro/pull/1225): if you hadn't edited in `RELEASE.md` then this PR would only be running the docs workflows on CircleCI. Possibly we should make that regex just look for .md files (rather than files just in docs/) so that PRs like this don't trigger all the code builds.

I noticed today when looking at the Windows/Python 3.6 errors that you all adopted dynamic config, nice!

File: docs/source/10_deployment/dask.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (AntonyMilneQB) [reply docs/source/10_deployment/dask.md:177]: Wow, great work searching the archives to find that! I wasn't even aware that `run_only_missing` existed until a couple of months ago when Nikos mentioned it.

**Comment by datajoely** [comment]
Well done @deepayman :pray:
